### PR TITLE
feat(Password): Allow special characters in password

### LINKED ISSUE #1049: Allow a set of special characters in passwords
## Description
- Allow passwords to contain some special characters: !@#$%^&*.
- Update form validation warning text to reflect new allowed characters.

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes

- Allow these symbols to be used in a password !@#$%^&*
- Passwords now require a letter, number, symbol, and to be at least 8 characters (we no longer require uppercase letter but they may still be used)
- Added password strength indicator
- Added visual password requirement validator

## Test

- Test with https://github.com/WISE-Community/WISE-API/pull/233

Make sure setting a password in these locations still works and validates with the new requirements
- Create teacher account
- Create student account
- Forgot teacher password
- Forgot student password
- User change password
- Teacher change student password

Closes #1049

### CODE REVIEW

File: src/app/common/password-helper.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { FormGroup } from '@angular/forms';
+import { NewPasswordAndConfirmComponent } from '../password/new-password-and-confirm/new-password-and-confirm.component';
+import { PasswordErrors } from '../domain/password/password-errors';
+
+export function changePasswordError(
+  error: any,
```
Comment (hirokiterashima) [inline src/app/common/password-helper.ts]: type PasswordErrors?

File: src/app/forgot/student/forgot-student-password-change/forgot-student-password-change.component.ts [hunk 2]
```diff
@@ -63,19 +64,9 @@ export class ForgotStudentPasswordChangeComponent implements OnInit {
   }
 
   private changePasswordError(error: any): void {
```
Comment (hirokiterashima) [inline src/app/forgot/student/forgot-student-password-change/forgot-student-password-change.component.ts]: type PasswordErrors?

File: src/app/forgot/teacher/forgot-teacher-password-change/forgot-teacher-password-change.component.ts [hunk 3]
```diff
@@ -68,7 +69,6 @@ export class ForgotTeacherPasswordChangeComponent implements OnInit {
   }
 
   private changePasswordError(error: any): void {
```
Comment (hirokiterashima) [inline src/app/forgot/teacher/forgot-teacher-password-change/forgot-teacher-password-change.component.ts]: type PasswordErrors?

File: src/app/modules/shared/edit-password/edit-password.component.spec.ts [hunk 4]
```diff
@@ -213,16 +220,28 @@ function setPasswords(oldPass: string, newPass: string, newPassConfirm: string)
...
+
+function generateErrorObservable(arg: string | any): Observable<any> {
+  return new Observable((observer) => {
+    observer.error(generateErrorResponseValue(arg));
+    observer.complete();
+  });
+}
+
+function generateErrorResponseValue(arg: string | any): any {
+  return typeof arg === 'string' ? { error: { messageCode: arg } } : { error: arg };
```
Comment (hirokiterashima) [inline src/app/modules/shared/edit-password/edit-password.component.spec.ts]: Reduce duplication? Do you need generateSuccessObservable() and generateErrorObservable(), or can we do it with a ternary in generateObservableResponse()?

```
function generateObservableResponse(arg: string | any, isSuccess: boolean): Observable<any> {
  return new Observable((observer) => {
    observer.next(isSuccess ? generateSuccessResponseValue(arg) : generateErrorResponseValue(arg));
    observer.complete();
  });
}
```

File: src/app/password/new-password-and-confirm/new-password-and-confirm.component.spec.ts [hunk 5]
```diff
@@ -1,119 +1,139 @@
...
+      expectedErrors: new PasswordErrors(false, false, false, true)
+    }
+  ];
+  errorCases.forEach(({ descriptionText, password, expectedErrors }) => {
+    describe(`password is ${descriptionText}`, () => {
+      it(`should indicate password is ${descriptionText}`, async () => {
+        await setPasswordAndExpectErrors(password, expectedErrors);
+      });
+    });
+  });
```
Comment (hirokiterashima) [inline src/app/password/new-password-and-confirm/new-password-and-confirm.component.spec.ts]: Change block to follow Arrange-Act-Assert pattern (https://github.com/CareMessagePlatform/jasmine-styleguide#arrange-act-assert)
Add beforeEach() to call the action (```await newPasswordAndConfirmHarness.setNewPassword(password);```), and then check the state inside the it() (```await checkPasswordRequirements(passwordErrors);```).

File: src/app/password/new-password-and-confirm/new-password-and-confirm.harness.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+  async isMissingSymbol(): Promise<boolean> {
+    return this.isMissingRequirement('symbol');
+  }
+
+  async isTooShort(): Promise<boolean> {
+    return this.isMissingRequirement('length');
+  }
+
+  async isMissingRequirement(
```
Comment (hirokiterashima) [inline src/app/password/new-password-and-confirm/new-password-and-confirm.harness.ts]: private?

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
Feature works great. I have some comments/questions:

- Do we want to require symbols? This would make passwords more difficult to remember, especially students.
- Always showing password strength indicator by default when the pages load feels a bit off because I haven't typed in anything yet. Should it show up once you start typing a password?
- And also maybe @breity can style the indicator a bit to fit into the page better and ensure contrast (a11y)?
- Change success message from "Password Changed" to "Successfully changed password" to make it more friendly?
- Code improvements
   - See inline comments
   - See if you can remove all duplications? 
   - For the new tests in this change set
      - change it('should disable ...') to it('disables ...')  https://github.com/CareMessagePlatform/jasmine-styleguide#speak-human?
      - try to change test blocks to follow Arrange-Act-Assert pattern (https://github.com/CareMessagePlatform/jasmine-styleguide#arrange-act-assert) ?

**Review by hirokiterashima — COMMENTED** [review]
Can we try not to introduce these new duplications?
1. generateSuccessObservable() and generateErrorObservable()
2. setNewPassword() and setConfirmNewPassword()

**Review by hirokiterashima — COMMENTED** [review]
New changes look good.

We still need to discuss/address these:
- Always showing password strength indicator by default when the pages load feels a bit off because I haven't typed in anything yet. Should it show up once you start typing a password?
- Maybe @breity can style the indicator a bit to fit with the rest of the page better and ensure that a11y standards are met (e.g. has good contrast)?

**Comment by geoffreykwan** [comment]
@breity I needed to add this to new-password-and-confirm.component.scss in order to set the width of the menu since the default width caused some text to wrap. Is there a better way to do this without ng-deep?

```
::ng-deep .mat-mdc-menu-panel {
  max-width: none !important;
}
```

**Comment by breity** [comment]
I removed the ng-deep requirement by setting view encapsulation to none for the component. Just need to be careful to make sure component styles don't modify any styles used elsewhere in the app when doing so.

I also cleaned up the styles/layout a bit and addressed text color contrast for a11y. I removed the changing color for the text indicating password strength ("Very Weak", "Weak", ...) because it's difficult to find colors with sufficient contrast in the yellow and orange ranges unless they're rather dark. I think the colors on the strength meter combined with text itself works fine.

**Review by hirokiterashima — APPROVED** [review]
New popup behavior looks good.

Can you add margin or padding to the text in the popup? Everything looks very close to the edge of the popup, which makes it a bit hard to read:
image

**Pull request merged**